// node-plop/lib/actions/addMany.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _co = require('co');

var _co2 = _interopRequireDefault(_co);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _fsPromiseProxy = require('../fs-promise-proxy');

var _globby = require('globby');

var _globby2 = _interopRequireDefault(_globby);

var _commonActionInterfaceCheck = require('./_common-action-interface-check');

var _commonActionInterfaceCheck2 = _interopRequireDefault(_commonActionInterfaceCheck);

var _commonActionAddFile = require('./_common-action-add-file');

var _commonActionAddFile2 = _interopRequireDefault(_commonActionAddFile);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = _co2.default.wrap(function* (data, cfg, plop) {
	// check the common action interface attributes. skip path check because it's NA
	const interfaceTestResult = (0, _commonActionInterfaceCheck2.default)(cfg, { checkPath: false });
	if (interfaceTestResult !== true) {
		throw interfaceTestResult;
	}
	// check that destination (instead of path) is a string value
	const dest = cfg.destination;
	if (typeof dest !== 'string' || dest.length === 0) {
		throw `Invalid destination "${dest}"`;
	}

	if (cfg.base) {
		cfg.base = plop.renderString(cfg.base, data);
	}

	if (typeof cfg.templateFiles === 'function') {
		cfg.templateFiles = cfg.templateFiles();
	}

	cfg.templateFiles = [].concat(cfg.templateFiles) // Ensure `cfg.templateFiles` is an array, even if a string is passed.
	.map(file => plop.renderString(file, data)); // render the paths as hbs templates

	const templateFiles = resolveTemplateFiles(cfg.templateFiles, cfg.base, cfg.globOptions, plop);

	const filesAdded = [];
	for (let templateFile of templateFiles) {
		const fileCfg = Object.assign({}, cfg, {
			path: resolvePath(cfg.destination, templateFile, cfg.base),
			template: yield (0, _fsPromiseProxy.readFile)(_path2.default.resolve(plop.getPlopfilePath(), templateFile))
		});
		const addedPath = yield (0, _commonActionAddFile2.default)(data, fileCfg, plop);

		filesAdded.push(addedPath);
	}

	return `${filesAdded.length} files added\n -> ${filesAdded.join('\n -> ')}`;
});


function resolveTemplateFiles(templateFilesGlob, basePath, globOptions, plop) {
	globOptions = Object.assign({ cwd: plop.getPlopfilePath() }, globOptions);
	return _globby2.default.sync(templateFilesGlob, Object.assign({ nobrace: true }, globOptions)).filter(isUnder(basePath)).filter(isAbsoluteOrRelativeFileTo(plop.getPlopfilePath()));
}
function isAbsoluteOrRelativeFileTo(relativePath) {
	const isFile = file => _fs2.default.existsSync(file) && _fs2.default.lstatSync(file).isFile();
	return file => isFile(file) || isFile(_path2.default.join(relativePath, file));
}

function isUnder() {
	let basePath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

	return path => path.startsWith(basePath);
}

function resolvePath(destination, file, rootPath) {
	return toUnix(_path2.default.join(destination, dropFileRootPath(file, rootPath)));
}

function toUnix(path) {
	return !path.sep || path.sep === '\\' ? path.replace(/\\/g, '/') : path;
}

function dropFileRootPath(file, rootPath) {
	return rootPath ? file.replace(rootPath, '') : dropFileRootFolder(file);
}

function dropFileRootFolder(file) {
	const fileParts = _path2.default.normalize(file).split(_path2.default.sep);
	fileParts.shift();

	return fileParts.join(_path2.default.sep);
}
