// node-plop/lib/generator-runner.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

exports.default = function (plopfileApi, flags) {
	let abort;

	// triggers inquirer with the correct prompts for this generator
	// returns a promise that resolves with the user's answers
	const runGeneratorPrompts = _co2.default.wrap(function* (genObject) {
		let bypassArr = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
		const prompts = genObject.prompts;


		if (prompts == null) {
			throw Error(`${genObject.name} has no prompts`);
		}

		if (typeof prompts === 'function') {
			return yield prompts(plopfileApi.inquirer);
		}

		// handle bypass data when provided

		var _promptBypass = (0, _promptBypass4.default)(prompts, bypassArr, plopfileApi),
		    _promptBypass2 = _slicedToArray(_promptBypass, 2);

		const promptsAfterBypass = _promptBypass2[0],
		      bypassAnswers = _promptBypass2[1];


		return yield plopfileApi.inquirer.prompt(promptsAfterBypass).then(answers => Object.assign(answers, bypassAnswers));
	});

	// Run the actions for this generator
	const runGeneratorActions = _co2.default.wrap(function* (genObject, data) {
		var changes = []; // array of changed made by the actions
		var failures = []; // array of actions that failed
		var actions = genObject.actions; // the list of actions to execute

		const customActionTypes = getCustomActionTypes();
		const actionTypes = Object.assign({}, customActionTypes, buildInActions);

		abort = false;

		// if action is a function, run it to get our array of actions
		if (typeof actions === 'function') {
			actions = actions(data);
		}

		// if actions are not defined... we cannot proceed.
		if (actions == null) {
			throw Error(`${genObject.name} has no actions`);
		}

		// if actions are not an array, invalid!
		if (!(actions instanceof Array)) {
			throw Error(`${genObject.name} has invalid actions`);
		}

		for (let _ref of actions.entries()) {
			var _ref2 = _slicedToArray(_ref, 2);

			let actionIdx = _ref2[0];
			let action = _ref2[1];

			// bail out if a previous action aborted
			if (abort) {
				failures.push({
					type: action.type || '',
					path: action.path || '',
					error: 'Aborted due to previous action failure'
				});
				continue;
			}

			action.force = flags.force === true || action.force === true;

			const actionIsFunction = typeof action === 'function';
			const actionCfg = actionIsFunction ? {} : action;
			const actionLogic = actionIsFunction ? action : actionTypes[actionCfg.type];

			if (typeof actionLogic !== 'function') {
				if (actionCfg.abortOnFail !== false) {
					abort = true;
				}
				failures.push({
					type: action.type || '',
					path: action.path || '',
					error: `Invalid action (#${actionIdx + 1})`
				});
				continue;
			}

			try {
				const actionResult = yield executeActionLogic(actionLogic, actionCfg, data);
				changes.push(actionResult);
			} catch (failure) {
				failures.push(failure);
			}
		}

		return { changes: changes, failures: failures };
	});

	// handle action logic
	const executeActionLogic = _co2.default.wrap(function* (action, cfg, data) {
		const failure = makeErrorLogger(cfg.type || 'function', '', cfg.abortOnFail);

		// convert any returned data into a promise to
		// return and wait on
		const fullData = Object.assign({}, cfg.data, data);
		return yield Promise.resolve(action(fullData, cfg, plopfileApi)).then(
		// show the resolved value in the console
		result => ({
			type: cfg.type || 'function',
			path: _colors2.default.blue(result.toString())
		}),
		// a rejected promise is treated as a failure
		function (err) {
			throw failure(err.message || err.toString());
		});
	});

	// request the list of custom actions from the plopfile
	function getCustomActionTypes() {
		return plopfileApi.getActionTypeList().reduce(function (types, name) {
			types[name] = plopfileApi.getActionType(name);
			return types;
		}, {});
	}

	// provide a function to handle action errors in a uniform way
	function makeErrorLogger(type, path, abortOnFail) {
		return function (error) {
			if (abortOnFail !== false) {
				abort = true;
			}
			return { type: type, path: path, error: error };
		};
	}

	return {
		runGeneratorActions: runGeneratorActions,
		runGeneratorPrompts: runGeneratorPrompts
	};
};

var _co = require('co');

var _co2 = _interopRequireDefault(_co);

var _colors = require('colors');

var _colors2 = _interopRequireDefault(_colors);

var _promptBypass3 = require('./prompt-bypass');

var _promptBypass4 = _interopRequireDefault(_promptBypass3);

var _actions = require('./actions');

var buildInActions = _interopRequireWildcard(_actions);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
